libquery/src/CypherCompiler.cpp

```
@@ -527,103 +382,47 @@ int galois::CypherCompiler::CompilePatternPath(CypherASTNode ast, bool negate) {
   }
 }
 
-galois::CompilerTreeVariant
+std::unique_ptr<galois::Expression>
 galois::CypherCompiler::CompileBinaryOperator(CypherASTNode ast) {
   CypherASTOp op = cypher_ast_binary_operator_get_operator(ast);
-  std::variant<ExpressionTreeOperator, ExpressionClauseOperator,
-               ArithmeticTreeOperator>
-      clause_op;
-
+  ExpressionOperator exp_op;
   // Find the type of the op and save it to a variant that is used to determine
   // how to connect 2 subtrees
   if (op == CYPHER_OP_AND) {
-    clause_op.emplace<kExpressionTree>(kTreeAnd);
+    exp_op = kAnd;
   } else if (op == CYPHER_OP_OR) {
-    clause_op.emplace<kExpressionTree>(kTreeOr);
+    exp_op = kOr;
   } else if (op == CYPHER_OP_XOR) {
-    clause_op.emplace<kExpressionTree>(kTreeXor);
+    exp_op = kXor;
   } else if (op == CYPHER_OP_EQUAL) {
-    clause_op.emplace<kExpressionClause>(kClauseEqual);
+    exp_op = kEqual;
   } else if (op == CYPHER_OP_NEQUAL) {
-    clause_op.emplace<kExpressionClause>(kClauseNotEqual);
+    exp_op = kNEqual;
   } else if (op == CYPHER_OP_PLUS) {
-    clause_op.emplace<kArithmeticTree>(kArithAdd);
+    exp_op = kAdd;
   } else if (op == CYPHER_OP_MINUS) {
-    clause_op.emplace<kArithmeticTree>(kArithSubtract);
+    exp_op = kSubtract;
   } else if (op == CYPHER_OP_MULT) {
-    clause_op.emplace<kArithmeticTree>(kArithMultiply);
+    exp_op = kMultiply;
   } else if (op == CYPHER_OP_DIV) {
-    clause_op.emplace<kArithmeticTree>(kArithDivide);
+    exp_op = kDivide;
   } else if (op == CYPHER_OP_MOD) {
-    clause_op.emplace<kArithmeticTree>(kArithModulo);
+    exp_op = kModulo;
   } else {
-    // return empty
-    GALOIS_LOG_WARN(
-        "Unsupported binary operation; undefined behavior may occur");
-    return CompilerTreeVariant();
+    GALOIS_LOG_FATAL("Unsupported binary operation");
```

guess you're going for "fail fast" now

---

Sometimes this is called the samurai principle: succeed or never return. I actually don't advocate it in general, but it has a better name than expose errors thoughtfully.